dsgrid/config/project_config.py

```diff
@@ -730,46 +781,204 @@ def _iter_datasets_by_status(self, status: DatasetRegistryStatus) -> InputDatase
+                dim_type,
+                supp_dim_query_names,
+            )
+            single_dfs[field] = list(record_ids)
+        single_df = create_dataframe_from_product(single_dfs)
+        return cross_join_dfs(multi_dfs + [single_df])
+
+    def _build_multi_dim_requirement_associations(
+        self, multi_dim_reqs, pivoted_dimension, supp_dim_query_names
+    ):
+        dfs_by_dim_combo = {}
+
+        # Example: Partial sector and subsector combinations are required.
+        # [
+        #     {"sector": "com", "subsector": ["commercial_subsector_supplemental"]},
+        #     {"sector": "res", "subsector": ["MidriseApartment"]},
+        # ]
+        # This code will replace supplemental records with base records and return a list of
+        # dataframes of those combinations - one per unique combination of dimensions.
+        for item in multi_dim_reqs:
+            dim_combo = []
+            columns = {}
+            for field in RequiredDimensionRecordsModel.__fields__:
+                record_ids = set(getattr(item, field))
+                if record_ids:
+                    dim_type = DimensionType(field)
+                    if dim_type == pivoted_dimension:
+                        raise NotImplementedError(
+                            f"Unhandled condition: multi_dimensional requirement cannot contain "
+                            f"the pivoted dimension: dimension={dim_type} records={record_ids}"
+                        )
+                    record_ids = self._replace_supplemental_record_ids_with_base(
+                        record_ids,
+                        dim_type,
+                        supp_dim_query_names,
+                    )
+                    dim_combo.append(dim_type.value)
+                    columns[field] = list(record_ids)
+            df = create_dataframe_from_product(columns)
+            df = df.select(*sorted(df.columns))
+
+            dim_combo = tuple(sorted(dim_combo))
+            if dim_combo in dfs_by_dim_combo:
+                dfs_by_dim_combo[dim_combo] = dfs_by_dim_combo[dim_combo].union(df)
+            else:
+                dfs_by_dim_combo[dim_combo] = df
+
+        return list(dfs_by_dim_combo.values())
+
+    def _replace_supplemental_record_ids_with_base(
+        self, record_ids, dim_type, supp_dim_query_names
+    ):
+        base_record_ids = self.get_base_dimension(dim_type).get_unique_ids()
+        if not record_ids:
+            return base_record_ids
+
+        supplemental_ids = record_ids.difference(base_record_ids)
+        for supplemental_id in supplemental_ids:
+            new_base_record_ids = self._map_supplemental_record_to_base_records(
+                supp_dim_query_names,
+                dim_type,
+                supplemental_id,
+            )
+            record_ids.remove(supplemental_id)
+            record_ids.update(new_base_record_ids)
+
+        return record_ids
+
+    def _map_supplemental_record_to_base_records(
```

Comment:
one record is singular the other is plural?

Reply:
This function replaces one supplemental dimension record ID with its corresponding base dimension record IDs.